consensus/istanbul/core/core.go

```diff
@@ -263,52 +304,47 @@ func (c *core) startNewRound(round *big.Int) {
 	c.updateRoundState(newView, c.valSet, roundChange)
 	// Calculate new proposer
 	c.valSet.CalcProposer(lastProposer, newView.Round.Uint64())
-	c.waitingForRoundChange = false
 	c.setState(StateAcceptRequest)
-	if roundChange && c.isProposer() && c.current != nil {
-		// If it is locked, propose the old proposal
-		// If we have pending request, propose pending request
-		if c.current.IsHashLocked() {
-			r := &istanbul.Request{
-				Proposal: c.current.Proposal(), //c.current.Proposal would be the locked proposal by previous proposer, see updateRoundState
-			}
-			c.sendPreprepare(r)
-		} else if c.current.pendingRequest != nil {
-			c.sendPreprepare(c.current.pendingRequest)
-		}
+	if roundChange && c.isProposer() && c.current != nil && request != nil {
+		c.sendPreprepare(request, roundChangeCertificate)
 	}
 	c.newRoundChangeTimer()
 
 	logger.Debug("New round", "new_round", newView.Round, "new_seq", newView.Sequence, "new_proposer", c.valSet.GetProposer(), "valSet", c.valSet.List(), "size", c.valSet.Size(), "isProposer", c.isProposer())
 }
 
-func (c *core) catchUpRound(view *istanbul.View) {
-	logger := c.logger.New("old_round", c.current.Round(), "old_seq", c.current.Sequence(), "old_proposer", c.valSet.GetProposer())
-
-	if view.Round.Cmp(c.current.Round()) > 0 {
-		c.roundMeter.Mark(new(big.Int).Sub(view.Round, c.current.Round()).Int64())
+// All actions that occur when transitioning to waiting for round change state.
+func (c *core) waitForDesiredRound(r *big.Int) {
+	logger := c.logger.New("func", "waitForDesiredRound", "cur_round", c.current.Round(), "old_desired_round", c.current.DesiredRound(), "new_desired_round", r)
+	// Don't wait for an older round
+	if c.current.DesiredRound().Cmp(r) >= 0 {
+		logger.Debug("New desired round not greater than current desired round")
+		return
 	}
-	c.waitingForRoundChange = true
-
-	// Need to keep block locked for round catching up
-	c.updateRoundState(view, c.valSet, true)
-	c.roundChangeSet.Clear(view.Round)
-	c.newRoundChangeTimer()
+	logger.Debug("Waiting for desired round")
 
-	logger.Trace("Catch up round", "new_round", view.Round, "new_seq", view.Sequence, "new_proposer", c.valSet)
+	desiredView := &istanbul.View{
+		Sequence: new(big.Int).Set(c.current.Sequence()),
+		Round:    new(big.Int).Set(r),
+	}
+	// Perform all of the updates
+	c.setState(StateWaitingForNewRound)
```

Comment:
There's two components to this state: `desiredRound` and `WaitingForNewRound`. Because we can timeout in `StateWaitingForNewRound`, we should not bail out if we're already in the waiting state, but **should** bail if the given desired round is not greater than new desired round.

`StateWaitingForNewRound` is pretty much only used to reject incoming messages.